### ISSUE TITLE
Error for duplicates in union

### ISSUE THREAD

**Issue by Akuli** [microsoft/pyright#2753]
This isn't really important, but it would be a nice touch.

**Is your feature request related to a problem? Please describe.**

`Foo = str | str` is not an error.

```
akuli@akuli-desktop:~/typeshed$ cat asd.py
def foo(thing: str | str) -> None: ...
def bar(thing: str | str | int) -> None: ...

akuli@akuli-desktop:~/typeshed$ npx -p pyright@1.1.192 pyright --pythonversion 3.11 asd.py 
npx: installed 1 in 1.714s
Loading configuration file at /home/akuli/typeshed/pyrightconfig.json
Assuming Python platform Linux
stubPath /home/akuli/typeshed/typings is not a valid directory.
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 infos 
Completed in 0.798sec
```

**Describe the solution you'd like**
Error when a union has the same AST in two members. Checking if it is the same type isn't great, because it would cause false positives in code like this:

```
if sys.version >= (3, 0):
    _Text = str
else:
    _Text = unicode

def foo(thing: str | _Text) -> None: ...
```

**Additional context**
Typeshed contained code that has duplicate union members as a result of find/replacing: https://github.com/python/typeshed/pull/6652

**Comment by erictraut** [comment]
Unions are allowed to contain duplicates, which are elided in the final type. I don't think this case is problematic enough or a common enough source of bugs to justify adding a specific check for it. Mypy doesn't generate an error in this case. The TypeScript compiler (which has a similar type system) also does not generate an error in this case.

Adding this check for the new union syntax would actually be a bit of a challenge because the meaning of `|` is ambiguous at parsing time. That means the check would need to be implemented at the time of semantic (type) evaluation, and the type checking logic would need to grovel deep in the parse tree to detect this case. It's theoretically possible, but it would be pretty gross.

I'm going to close this for now. If this issue gets sufficient upvotes (thumbs up), we might reconsider this decision.

**Comment by JelleZijlstra** [comment]
We could implement this in https://github.com/ambv/flake8-pyi/ pretty easily, and because all instances of `|` in stubs mean Union, it would not be as complicated as it would be in pyright.
